src/rest/SearchEndpoint.ts

```diff
@@ -873,4 +821,404 @@ export class SearchEndpoint implements ISearchEndpoint {
   private hasTrailingSlash(uri: string) {
     return uri.charAt(uri.length - 1) == '/';
   }
+
+  private isMissingAuthenticationProviderStatus(status: number): boolean {
+    return status == 402;
+  }
+
+  private isAccessTokenExpiredStatus(status: number): boolean {
+    return status == 419;
+  }
 }
+
+
+/**
+ * It's taken for granted that methods using decorators have :
+ * {@link IEndpointCallOptions} as their second to last parameter
+ * {@link IEndpointCallParameters} as their last parameter
+ * The default parameters for each member of the injected {@link IEndpointCallParameters} are the following:
+ * url: '',
+ * queryString: [],
+ * requestData: {},
+ * requestDataType: undefined,
+ * method: '',
+ * responseType: '',
+ * errorsAsSuccess: false
+ */
+
+/**
+ * Add the base url
+ * @param path The path to append to the url
+ */
+function url(path: string){
+  return function(target: Object, key: string, descriptor: TypedPropertyDescriptor<any>) {
+    let originalMethod = descriptor.value;
+    let nbParams = target[key].prototype.constructor.length;
+
+    descriptor.value = function(...args: any[]) {
+      let uri = this.buildBaseUri(path);
+      if(args[nbParams-1]){
+        args[nbParams-1].url = uri;
+      }
+      else{
+        let params: IEndpointCallParameters = {
+          url: uri,
+          queryString: [],
+          requestData: {},
+          method: '',
+          responseType: '',
+          errorsAsSuccess: false
+        };
+        args[nbParams-1] = params;
+      }
+      let result = originalMethod.apply(this, args);
+      return result;
+    };
+
+    return descriptor;
+  }
+}
+
+/**
+ * Add the alert url
+ * @param path The path to append to the url
+ */
+function alertsUrl(path: string){
+  return function(target: Object, key: string, descriptor: TypedPropertyDescriptor<any>) {
+    let originalMethod = descriptor.value;
+    let nbParams = target[key].prototype.constructor.length;
+
+    descriptor.value = function(...args: any[]) {
+      let uri = this.buildSearchAlertsUri(path);
+      if(args[nbParams-1]){
+        args[nbParams-1].url = uri;
+      }
+      else{
+        let params: IEndpointCallParameters = {
+          url: uri,
+          queryString: [],
+          requestData: {},
+          method: '',
+          responseType: '',
+          errorsAsSuccess: false
+        };
+        args[nbParams-1] = params;
+      }
+      let result = originalMethod.apply(this, args);
+      return result;
+    };
+
+    return descriptor;
+  }
+}
+
+/**
+ * Add the base query string arguments
+ */
+function queryStringArguments(){

```

Comment:
Since this is needed for every single call on the endpoint (correct me if i'm wrong), I think it should then not be passed as a decorator (which means it's not "mandatory")

I would instead move the logic for the base query strings arguments to somewhere where we are sure it will always be added.

Maybe the `performOneCall` method